### PR TITLE
GH-2280 Detect and report when reaching quota in local storage

### LINKED ISSUE #2280: [local storage] Detect and report when quota is reached
The browser's local storage is used by Theia to store the state of the workbench on exit. This permits e.g. reloading the client, and closing a workspace and later re-opening it, in both cases without losing the previous state.

The preserved data is saved in local storage, independently per origin - protocol, host and port (e.g. http://localhost:3000 is one origin). IIUC the local storage quota is per-origin and relatively low - ~5-10 MB depending on the browser. 

When Theia is used on a desktop, the origin will likely always be the same (if the same port is used), and so all workspaces saved state will share the same local storage / quota. ATM the problem markers are preserved as part of that state, and in some situation they can be numerous, eating at the quota that much faster. 

Eventually some users will fill the local storage for a given origin. ATM this case is not detected; Theia will just silently fail to save any more data. __We should at least detect when this happens and notify the user with basic information (e.g. link) about how to deal with this.__

A second step could be to provide a UI to help the user to selectively get rid of old information that might no longer be needed. We have a separate issue for that: #2272

### PR DESCRIPTION
**Pull request by lmcbout**
Fixes #2280 
Signed-off-by: Jacques Bouthillier <jacques.bouthillier@ericsson.com>

### CODE REVIEW

**Comment by lmcbout** [comment]
Rebase on top of issue #2357

File: packages/core/src/browser/storage-service.ts [hunk 1]
```diff
@@ -43,9 +44,12 @@ interface LocalStorage {
 @injectable()
 export class LocalStorageService implements StorageService {
     private storage: LocalStorage;
+    private fullstorage: boolean = false;
```
Comment (svenefftinge) [inline packages/core/src/browser/storage-service.ts]: why coping the state into a field?

File: packages/core/src/browser/storage-service.ts [hunk 2]
```diff
@@ -57,7 +61,17 @@ export class LocalStorageService implements StorageService {
...
-            this.storage[this.prefix(key)] = JSON.stringify(data);
+            try {
+                if (!this.isLocalStorageFull()) {
+                    this.storage[this.prefix(key)] = JSON.stringify(data);
+                }
+            } catch (e) {
+                if (this.messageService) {
+                    this.fullstorage = true;
+                    const currentSize = this.verifyLocalStorage();
+                    this.messageService.warn(`Web storage quota: \n  ${e} \n current size: ${currentSize} bytes`);
```
Comment (svenefftinge) [inline packages/core/src/browser/storage-service.ts]: I don't think this message helps the user, as it doesn't explain what it is about.
Also it might be that there are already many old buckets under the same URI. We might better ask the user whether we should clean 'old' data.

File: packages/core/src/browser/storage-service.ts [hunk 3]
```diff
@@ -76,4 +90,18 @@ export class LocalStorageService implements StorageService {
...
+
+        let totalLength = 0;
+        for (let count = 0; count < this.storage.length; count++) {
+            const key = this.storage.key(count);
+            totalLength += this.storage[key].length;
+        }
+        return totalLength.toString();
+    }
+
+    private isLocalStorageFull(): boolean {
```
Comment (svenefftinge) [inline packages/core/src/browser/storage-service.ts]: unnecessary method

File: packages/core/src/browser/storage-service.ts (hunk 1 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: We try to copy the data to the local storage only if there is still some room. Once we reach the quota, no need to try to write to the local storage anymore. The flag is to maintain the value if the user has previously reach the maximun.

File: packages/core/src/browser/storage-service.ts (hunk 2 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: This issue was just to detect when there is no more local storage availble to the end user. Before the issue fix by @akosyakov ( commit a5f78521a7b9ec2a190111877db70932f3f0c829 [markers] don't store by default ) offerred on July 10, it was really easy to reach the local storage quota when compiling Java project.
Should we put something like: "Proposed message: You web local storage is full < size>, you need to remove some of the data stored to allow new data to be saved. The new data could be for example you current working layout." until we fixed issue #2272 ?
The issue #2272 is to find a solution to let the end user decide what to do . Some solutions could be : 
- clear his web cache (not necessary the best solution)
- propose to the user a list of data already stored then let him decide what should be removed.
- clean up automatically the workspace that does not exist anymore

File: packages/core/src/browser/storage-service.ts (hunk 3 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Removed

File: packages/core/src/browser/storage-service.ts (hunk 1 above)
Comment (marcdumais-work) [reply packages/core/src/browser/storage-service.ts]: If we keep this field, storageFull may be a better name

File: packages/core/src/browser/storage-service.ts [hunk 4]
```diff
@@ -57,7 +61,17 @@ export class LocalStorageService implements StorageService {
 
     setData<T>(key: string, data?: T): Promise<void> {
         if (data !== undefined) {
-            this.storage[this.prefix(key)] = JSON.stringify(data);
+            try {
+                if (!this.fullstorage) {
```
Comment (marcdumais-work) [inline packages/core/src/browser/storage-service.ts]: ATM there is no way for ```fullstorage``` to go back to a ```false``` value, until the client is restarted. If e.g. the user cleans-up the local storage without restarting the client, we would still skip attempting to save in the local storage.

Instead we might try to save to local storage every time, but only report that we failed once.

**Comment by marcdumais-work** [comment]
@lmcbout There seems to be a React-related issue on Travis. I do not see that issue on GitPod. Maybe rebasing this PR would help?

**Comment by lmcbout** [comment]
Modify the solution, so we check if there are still some room available in the local storage 
Testing with a 300K buffer (Size to cover a workspace layout with some views, some files open in the editor, outline view problem view opened)
If the test does not succeed, the pop up show a message and point to a web page which will explain how to clean up some local storage memory.
To test, just increase the buffer array (in storage-service.testLocalStorage() to a size of 1.2 M), it will trigger the quota exceeed message
The message:
            Your prefered browser local storage is almost full.
            To be able to save your current workspace layout or data, you may need to free some space.
            You can refer to Theia page for instructions.
            on how to manually clean your browser local storage

And the link to explain how to clean your local storage: 
https://github.com/theia-ide/theia/wiki/Cleaning-Local-Storage

File: packages/core/src/browser/storage-service.ts [hunk 5]
```diff
@@ -76,4 +85,33 @@ export class LocalStorageService implements StorageService {
...
+        const array = new Array(60000); // size: <array size> * 5 =  ~ 300K
+        const keyTest = this.prefix(`Test`);
+
+        if (this.storage !== this.fallbackStorage) {
+            try {
+                this.storage[keyTest] = JSON.stringify(array);
+                this.storage.removeItem(keyTest);
+            } catch (error) {
+                this.ShowMessageNotification();
+                this.storage.removeItem(keyTest);
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: perhaps use a ``try catch finally`` since in both cases we will need to make sure that the test data is removed and we want to minimize duplicate code?

Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: perhaps use a ``try catch finally`` since in both cases we will need to make sure that the test data is removed and we want to minimize duplicate code?

Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Done, Use finally

**Comment by lmcbout** [comment]
Rebase on the lastest as of Aug 28, 2018
Here the message dialog that shows when the user has less than 300K memory left on the browser to save his workspace configuration:
image

File: packages/core/src/browser/storage-service.ts [hunk 6]
```diff
@@ -76,4 +84,31 @@ export class LocalStorageService implements StorageService {
         const pathname = typeof window === 'undefined' ? '' : window.location.pathname;
         return `theia:${pathname}:${key}`;
     }
+
+    private ShowMessageNotification() {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: Small typo with the use of incorrect `PascalCase` method name. Also, perhaps rename the method to something less generic such as `showDiskQuotaExceededMessage`/`showStorageExceededMessage` or along similar lines.

File: packages/core/src/browser/storage-service.ts [hunk 7]
```diff
@@ -45,10 +46,13 @@ export class LocalStorageService implements StorageService {
     private storage: LocalStorage;
 
     constructor(
-        @inject(ILogger) protected logger: ILogger
+        @inject(ILogger) protected logger: ILogger,
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: Use `property injection` rather than `constructor injection` for new dependencies.

File: packages/core/src/browser/storage-service.ts [hunk 8]
```diff
@@ -76,4 +84,31 @@ export class LocalStorageService implements StorageService {
...
+            You can refer to Theia page for instructions.
+            on how to manually clean your browser local storage.`;
+        this.messageService.warn(ERROR_MESSAGE, READ_INSTRUCTIONS_ACTION).then(selected => {
+            if (READ_INSTRUCTIONS_ACTION === selected) {
+                window.open('https://github.com/theia-ide/theia/wiki/Cleaning-Local-Storage');
+            }
+        });
+    }
+
+    private testLocalStorage(): void {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: Please include documentation.

File: packages/core/src/browser/storage-service.ts [hunk 9]
```diff
@@ -76,4 +84,31 @@ export class LocalStorageService implements StorageService {
         const pathname = typeof window === 'undefined' ? '' : window.location.pathname;
         return `theia:${pathname}:${key}`;
     }
+
+    private ShowMessageNotification() {
+        const READ_INSTRUCTIONS_ACTION = 'Read Instructions';
+        const ERROR_MESSAGE = `Your prefered browser local storage is almost full.
+            To be able to save your current workspace layout or data, you may need to free some space.
+            You can refer to Theia page for instructions.
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: There are typos in your `ERROR_MESSAGE`. 
`Your preferred browser's local storage is almost full.
To be able to save your current workspace layout or data, you may need to free up some space.
You can refer to Theia's documentation page for instructions on how to manually clean your browser's local storage.`

File: packages/core/src/browser/storage-service.ts [hunk 10]
```diff
@@ -45,10 +46,13 @@ export class LocalStorageService implements StorageService {
     private storage: LocalStorage;
 
     constructor(
-        @inject(ILogger) protected logger: ILogger
+        @inject(ILogger) protected logger: ILogger,
+        @inject(MessageService) protected readonly messageService: MessageService,
+
```
Comment (vince-fugnitto) [inline packages/core/src/browser/storage-service.ts]: Use property injection rather than constructor injection for new dependencies.

File: packages/core/src/browser/storage-service.ts (hunk 6 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Done , use showDiskQuotaExceededMessage

File: packages/core/src/browser/storage-service.ts (hunk 8 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Documentation added for this method.

File: packages/core/src/browser/storage-service.ts (hunk 9 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Done.

File: packages/core/src/browser/storage-service.ts (hunk 10 above)
Comment (lmcbout) [reply packages/core/src/browser/storage-service.ts]: Still using the constructor injection. If I use the property injection, the front-end does not start and report the messageServide as undefined, so kept it with the constructor's injection.

Comment (AlexTugarev) [reply packages/core/src/browser/storage-service.ts]: that's because properties are injected after constructor call. please use `@postConstruct protected init() { }` for initialization relying on injected dependencies.
